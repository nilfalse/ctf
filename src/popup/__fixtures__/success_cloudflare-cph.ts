export const state = 'success' as const;

export const request = {
  url: 'https://ilin.dk/',
  ip: '104.27.165.61',
  responseHeaders: [
    {
      name: 'cf-ray',
      value: '5d5d5d5d5d5d5d5d-CPH',
    },
    {
      name: 'cf-cache-status',
      value: 'REVALIDATED',
    },
    {
      name: 'cf-request-id',
      value: '00000004f10000000000000000000001',
    },
    {
      name: 'x-nf-request-id',
      value: 'b1b1b1b1-c9c9-4545-afaf-292929292929-33333333',
    },
    {
      name: 'server',
      value: 'cloudflare',
    },
  ],
};

export const matches = [
  {
    heuristic: 'cloudflare',
    score: 1,
    isoCountry: 'DK',
    isoRegion: 'DK-84',
    continent: 'EU',
    extra: { ray: 'CPH', cacheStatus: 'REVALIDATED' },
  },
  {
    heuristic: 'ip',
    score: 0.5,
    isoCountry: 'US',
    isoRegion: null,
    continent: 'NA',
    extra: null,
  },
] as const;
