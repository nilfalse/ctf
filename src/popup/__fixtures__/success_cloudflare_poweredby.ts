export const state = 'success';

export const request = {
  url: 'https://peakon.com/',
  ip: '104.26.7.76',
  responseHeaders: [
    { name: 'x-powered-by', value: 'WP Engine' },
    { name: 'x-cacheable', value: 'SHORT' },
    { name: 'x-cache', value: 'HIT: 8' },
    { name: 'x-cache-group', value: 'normal' },
    { name: 'x-orig-cache-control', value: 'max-age=0' },
    { name: 'cf-cache-status', value: 'DYNAMIC' },
    { name: 'cf-request-id', value: '050505050500000ebebebebeb0000001' },
    { name: 'server', value: 'cloudflare' },
    { name: 'cf-ray', value: '5e5e5e5e5e5e5e5e7-FRA' },
  ],
};

export const matches = [
  {
    heuristic: 'cloudflare',
    score: 1,
    isoCountry: 'DE',
    isoRegion: 'DE-HE',
    continent: 'EU',
    extra: { ray: 'FRA', cacheStatus: 'DYNAMIC' },
  },
  {
    heuristic: 'ip',
    score: 0.5,
    isoCountry: 'US',
    isoRegion: null,
    continent: 'NA',
    extra: null,
  },
];
