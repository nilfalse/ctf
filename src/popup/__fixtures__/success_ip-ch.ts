export const state = 'success' as const;

export const request = {
  url: 'https://protonmail.com/',
  ip: '185.70.41.35',
};

export const matches = [
  {
    heuristic: 'ip',
    score: 0.5,
    isoCountry: 'CH',
    isoRegion: null,
    continent: 'EU',
    extra: { registeredCountry: 'CH' },
  },
] as const;
