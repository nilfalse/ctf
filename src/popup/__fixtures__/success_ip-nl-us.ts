export const state = 'success';

export const request = {
  url: 'https://duckduckgo.com/',
  ip: '40.114.177.156',
  responseHeaders: [
    { name: 'status', value: '304' },
    { name: 'server', value: 'nginx' },
  ],
};

export const matches = [
  {
    heuristic: 'ip',
    score: 0.5,
    isoCountry: 'NL',
    isoRegion: null,
    continent: 'EU',
    extra: { registeredCountry: 'US' },
  },
];
