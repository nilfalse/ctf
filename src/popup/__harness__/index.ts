export * from './content';
export * as render from './render';
