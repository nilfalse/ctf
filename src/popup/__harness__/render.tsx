import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function popupDependencies(wrapper: ReactWrapper) {
  await act(async () => {
    await import('country-flag-emoji-json/json/flag-emojis-by-code.json');

    await new Promise((resolve) => setTimeout(resolve));

    wrapper.update();
  });
}
