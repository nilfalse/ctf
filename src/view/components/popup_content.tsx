import { hot } from 'react-hot-loader/root';
// eslint-disable-next-line import/order
import * as React from 'react';

import { usePopupContent } from '../hooks/popup_content';

import { Empty } from './empty';
import { HeaderPrimary } from './header_primary';
import { Traceroute } from './traceroute';

export const _PopupContent: React.FC = () => {
  const popupResponse = usePopupContent();

  if (popupResponse.state === 'success' && popupResponse.details) {
    return (
      <>
        <HeaderPrimary {...popupResponse.details} />
        <Traceroute {...popupResponse.details} />
      </>
    );
  } else {
    return <Empty />;
  }
};

export const PopupContent = hot(_PopupContent);
