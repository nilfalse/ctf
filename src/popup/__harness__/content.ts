import { Match } from '../../heuristics';
import {
  CountryRequest,
  CountryRequestParams,
} from '../../lib/country_request';
import * as popupContent from '../../view/hooks/popup_content';

type HarnessContentParam =
  | popupContent.PopupContentInit
  | {
      state: 'success';
      request?: CountryRequestParams;
      matches: ReadonlyArray<Match>;
    }
  | popupContent.PopupContentFailure;

export function content(content: HarnessContentParam) {
  const spy = jest.spyOn(popupContent, 'usePopupContent');
  let mockedContent: popupContent.PopupContent;

  if (content.state === 'success') {
    const request = content.request
      ? new CountryRequest(content.request)
      : null;
    const details = request
      ? {
          request,
          matches: content.matches,
        }
      : null;

    mockedContent = { ...content, details } as const;
  } else {
    mockedContent = content;
  }

  beforeEach(function () {
    spy.mockReturnValue(mockedContent);
  });

  afterEach(function () {
    spy.mockReset();
  });

  return spy;
}
