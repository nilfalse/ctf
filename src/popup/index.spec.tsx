import { mount, ReactWrapper } from 'enzyme';
import * as React from 'react';

import { PopupContent } from '../view/components/popup_content';
import { Traceroute } from '../view/components/traceroute';
import { CloudflareTrace } from '../view/components/traceroute/cloudflare';
import { IPTrace } from '../view/components/traceroute/ip';

import * as successCloudflareCPH from './__fixtures__/success_cloudflare-cph';
import * as successCloudflarePoweredBy from './__fixtures__/success_cloudflare_poweredby';
import * as successIPSwitzerland from './__fixtures__/success_ip-ch';
import * as successIPNetherlandsUS from './__fixtures__/success_ip-nl-us';
import * as harness from './__harness__';

describe('Popup', () => {
  describe('for a Switzerland-based website', () => {
    let popup: ReactWrapper;

    harness.content(successIPSwitzerland);

    beforeEach(async () => {
      popup = mount(<PopupContent />);

      return harness.render.popupDependencies(popup);
    });

    afterEach(() => {
      popup.unmount();
    });

    it('should render IP country info', () => {
      const iptrace = popup.find(IPTrace);

      expect(iptrace.text()).toContain('Switzerland');
      expect(iptrace.children()).toMatchSnapshot();
    });
  });

  describe('for a Netherlands-based website with legal entity in the US', () => {
    let popup: ReactWrapper;

    harness.content(successIPNetherlandsUS);

    beforeEach(async () => {
      popup = mount(<PopupContent />);

      return harness.render.popupDependencies(popup);
    });

    afterEach(() => {
      popup.unmount();
    });

    it('should render IP country info', () => {
      const iptrace = popup.find(IPTrace);

      expect(iptrace.text()).toContain('Netherlands');
      expect(iptrace.text()).toContain('United States of America');
      expect(iptrace.children()).toMatchSnapshot();
    });
  });

  describe('for a Cloudflare-protected website', () => {
    let popup: ReactWrapper;

    harness.content(successCloudflareCPH);

    beforeEach(async () => {
      popup = mount(<PopupContent />);

      return harness.render.popupDependencies(popup);
    });

    afterEach(() => {
      popup.unmount();
    });

    it('should render country info for Cloudflare IP', () => {
      const iptrace = popup.find(IPTrace);

      expect(iptrace.text()).toContain('United States of America');
      expect(iptrace.children()).toMatchSnapshot();
    });

    it('should show Cloudflare data center abbreviation', () => {
      const cloudflareTrace = popup.find(CloudflareTrace).children();

      expect(cloudflareTrace.text()).toContain('Denmark');
      expect(cloudflareTrace.text()).toContain('CPH');
      expect(cloudflareTrace).toMatchSnapshot();
    });
  });

  describe('for a website behind Cloudflare with Powered By info exposed', () => {
    let popup: ReactWrapper;

    harness.content(successCloudflarePoweredBy);

    beforeEach(async () => {
      popup = mount(<PopupContent />);

      return harness.render.popupDependencies(popup);
    });

    afterEach(() => {
      popup.unmount();
    });

    it('should render country info for Cloudflare IP', () => {
      const iptrace = popup.find(IPTrace);

      expect(iptrace.text()).toContain('United States of America');
      expect(iptrace.children()).toMatchSnapshot();
    });

    it('should render the header value to the user', () => {
      const iptrace = popup.find(IPTrace).children();

      expect(iptrace.text()).toContain('Powered By:');
      expect(iptrace.text()).toContain('WP Engine');
      expect(iptrace).toMatchSnapshot();
    });

    it('should show Cloudflare data center abbreviation', () => {
      const cloudflareTrace = popup.find(CloudflareTrace).children();

      expect(cloudflareTrace.text()).toContain('Germany');
      expect(cloudflareTrace.text()).toContain('FRA');
      expect(cloudflareTrace).toMatchSnapshot();
    });
  });
});
