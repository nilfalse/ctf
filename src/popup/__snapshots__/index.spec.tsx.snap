// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup for a Cloudflare-protected website should render country info for Cloudflare IP 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        104.27.165.61
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.27.165.61"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.27.165.61"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": null,
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag="ðŸ‡ºðŸ‡¸"
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "cf-ray",
            "value": "5d5d5d5d5d5d5d5d-CPH",
          },
          Object {
            "name": "cf-cache-status",
            "value": "REVALIDATED",
          },
          Object {
            "name": "cf-request-id",
            "value": "00000004f10000000000000000000001",
          },
          Object {
            "name": "x-nf-request-id",
            "value": "b1b1b1b1-c9c9-4545-afaf-292929292929-33333333",
          },
          Object {
            "name": "server",
            "value": "cloudflare",
          },
        ],
        "ip": "104.27.165.61",
        "url": "https://ilin.dk/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              cloudflare
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup for a Cloudflare-protected website should show Cloudflare data center abbreviation 1`] = `
<div
  className="cloudflare-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      Cloudflare anycast rerouted through
    </p>
  </Component>
  <div
    className="traceroute__nested"
  >
    <CountryItem
      countryCode="DK"
    >
      <Component
        className="traceroute__country-item traceroute__country-item_flagged"
        data-flag="ðŸ‡©ðŸ‡°"
        size="xs"
      >
        <p
          className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡©ðŸ‡°"
        >
          <strong>
            CPH
          </strong>
           data center in
          <br />
          <strong>
            Denmark
          </strong>
        </p>
      </Component>
    </CountryItem>
    <Component
      size="xs"
      variant="light"
    >
      <p
        className="typography typography_light typography_xs"
      >
        Cache Status was 
        <Component
          variant="mono"
        >
          <span
            className="typography typography_mono"
          >
            REVALIDATED
          </span>
        </Component>
         (
        <Component
          href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
          size="xs"
          target="_blank"
          title="What does \\"Cache status\\" mean?"
          variant="light"
        >
          <a
            className="typography typography_light typography_xs link"
            href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
            target="_blank"
            title="What does \\"Cache status\\" mean?"
          >
            ?
          </a>
        </Component>
        )
      </p>
    </Component>
  </div>
</div>
`;

exports[`Popup for a Netherlands-based website with legal entity in the US should render IP country info 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        40.114.177.156
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=40.114.177.156"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=40.114.177.156"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "EU",
        "extra": Object {
          "registeredCountry": "US",
        },
        "heuristic": "ip",
        "isoCountry": "NL",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="NL"
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡³ðŸ‡±"
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag="ðŸ‡³ðŸ‡±"
          >
            Country:
            <br />
            <strong>
              Netherlands
            </strong>
          </p>
        </Component>
      </CountryItem>
      <CountryItem
        countryCode="US"
        shouldRenderFlag={true}
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag="ðŸ‡ºðŸ‡¸"
          >
            Legal registration:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "status",
            "value": "304",
          },
          Object {
            "name": "server",
            "value": "nginx",
          },
        ],
        "ip": "40.114.177.156",
        "url": "https://duckduckgo.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              nginx
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup for a Switzerland-based website should render IP country info 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        185.70.41.35
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=185.70.41.35"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=185.70.41.35"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "EU",
        "extra": Object {
          "registeredCountry": "CH",
        },
        "heuristic": "ip",
        "isoCountry": "CH",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="CH"
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡¨ðŸ‡­"
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag="ðŸ‡¨ðŸ‡­"
          >
            Country:
            <br />
            <strong>
              Switzerland
            </strong>
          </p>
        </Component>
      </CountryItem>
      <CountryItem
        countryCode="CH"
        shouldRenderFlag={false}
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag={null}
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag={null}
          >
            Legal registration:
            <br />
            <strong>
              Switzerland
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": undefined,
        "ip": "185.70.41.35",
        "url": "https://protonmail.com/",
      }
    }
  />
</div>
`;

exports[`Popup for a website behind Cloudflare with Powered By info exposed should render country info for Cloudflare IP 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        104.26.7.76
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": null,
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag="ðŸ‡ºðŸ‡¸"
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "x-powered-by",
            "value": "WP Engine",
          },
          Object {
            "name": "x-cacheable",
            "value": "SHORT",
          },
          Object {
            "name": "x-cache",
            "value": "HIT: 8",
          },
          Object {
            "name": "x-cache-group",
            "value": "normal",
          },
          Object {
            "name": "x-orig-cache-control",
            "value": "max-age=0",
          },
          Object {
            "name": "cf-cache-status",
            "value": "DYNAMIC",
          },
          Object {
            "name": "cf-request-id",
            "value": "050505050500000ebebebebeb0000001",
          },
          Object {
            "name": "server",
            "value": "cloudflare",
          },
          Object {
            "name": "cf-ray",
            "value": "5e5e5e5e5e5e5e5e7-FRA",
          },
        ],
        "ip": "104.26.7.76",
        "url": "https://peakon.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              cloudflare
            </span>
          </Component>
        </p>
      </Component>
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Powered By: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              WP Engine
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup for a website behind Cloudflare with Powered By info exposed should render the header value to the user 1`] = `
<div
  className="ip-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      You connected to IP 
      <strong>
        104.26.7.76
      </strong>
      Â 
      <Component
        href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
        size="xs"
        target="_blank"
        variant="light"
      >
        <a
          className="typography typography_light typography_xs link"
          href="https://apps.db.ripe.net/db-web-ui/query?source=GRS&bflag=true&searchtext=104.26.7.76"
          target="_blank"
        >
          lookup
        </a>
      </Component>
    </p>
  </Component>
  <CountryInfo
    match={
      Object {
        "continent": "NA",
        "extra": null,
        "heuristic": "ip",
        "isoCountry": "US",
        "isoRegion": null,
        "score": 0.5,
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <CountryItem
        countryCode="US"
      >
        <Component
          className="traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡ºðŸ‡¸"
          size="xs"
        >
          <p
            className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
            data-flag="ðŸ‡ºðŸ‡¸"
          >
            Country:
            <br />
            <strong>
              United States of America
            </strong>
          </p>
        </Component>
      </CountryItem>
    </div>
  </CountryInfo>
  <ServerInfo
    request={
      Object {
        "headers": Array [
          Object {
            "name": "x-powered-by",
            "value": "WP Engine",
          },
          Object {
            "name": "x-cacheable",
            "value": "SHORT",
          },
          Object {
            "name": "x-cache",
            "value": "HIT: 8",
          },
          Object {
            "name": "x-cache-group",
            "value": "normal",
          },
          Object {
            "name": "x-orig-cache-control",
            "value": "max-age=0",
          },
          Object {
            "name": "cf-cache-status",
            "value": "DYNAMIC",
          },
          Object {
            "name": "cf-request-id",
            "value": "050505050500000ebebebebeb0000001",
          },
          Object {
            "name": "server",
            "value": "cloudflare",
          },
          Object {
            "name": "cf-ray",
            "value": "5e5e5e5e5e5e5e5e7-FRA",
          },
        ],
        "ip": "104.26.7.76",
        "url": "https://peakon.com/",
      }
    }
  >
    <div
      className="traceroute__nested"
    >
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Server Software: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              cloudflare
            </span>
          </Component>
        </p>
      </Component>
      <Component
        size="xs"
      >
        <p
          className="typography typography_xs"
        >
          Powered By: 
          <Component
            variant="mono"
          >
            <span
              className="typography typography_mono"
            >
              WP Engine
            </span>
          </Component>
        </p>
      </Component>
    </div>
  </ServerInfo>
</div>
`;

exports[`Popup for a website behind Cloudflare with Powered By info exposed should show Cloudflare data center abbreviation 1`] = `
<div
  className="cloudflare-trace"
>
  <Component
    size="s"
  >
    <p
      className="typography typography_s"
    >
      Cloudflare anycast rerouted through
    </p>
  </Component>
  <div
    className="traceroute__nested"
  >
    <CountryItem
      countryCode="DE"
    >
      <Component
        className="traceroute__country-item traceroute__country-item_flagged"
        data-flag="ðŸ‡©ðŸ‡ª"
        size="xs"
      >
        <p
          className="typography typography_xs traceroute__country-item traceroute__country-item_flagged"
          data-flag="ðŸ‡©ðŸ‡ª"
        >
          <strong>
            FRA
          </strong>
           data center in
          <br />
          <strong>
            Germany
          </strong>
        </p>
      </Component>
    </CountryItem>
    <Component
      size="xs"
      variant="light"
    >
      <p
        className="typography typography_light typography_xs"
      >
        Cache Status was 
        <Component
          variant="mono"
        >
          <span
            className="typography typography_mono"
          >
            DYNAMIC
          </span>
        </Component>
         (
        <Component
          href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
          size="xs"
          target="_blank"
          title="What does \\"Cache status\\" mean?"
          variant="light"
        >
          <a
            className="typography typography_light typography_xs link"
            href="https://support.cloudflare.com/hc/articles/200172516-Understanding-Cloudflare-s-CDN#h_bd959d6a-39c0-4786-9bcd-6e6504dcdb97"
            target="_blank"
            title="What does \\"Cache status\\" mean?"
          >
            ?
          </a>
        </Component>
        )
      </p>
    </Component>
  </div>
</div>
`;
